File: pkg/resource_model/resource.go
package resource_model

import "github.com/google/uuid"
import "time"
import "github.com/apibrew/apibrew/pkg/formats/unstructured"

type Resource struct {
	Id          *uuid.UUID         `json:"id,omitempty"`
	Version     int32              `json:"version,omitempty"`
	CreatedBy   *string            `json:"createdBy,omitempty"`
	UpdatedBy   *string            `json:"updatedBy,omitempty"`
	CreatedOn   *time.Time         `json:"createdOn,omitempty"`
	UpdatedOn   *time.Time         `json:"updatedOn,omitempty"`
	Name        string             `json:"name,omitempty"`
	Namespace   *Namespace         `json:"namespace,omitempty"`
	Virtual     bool               `json:"virtual,omitempty"`
	Properties  []ResourceProperty `json:"properties,omitempty"`
	Indexes     []ResourceIndex    `json:"indexes,omitempty"`
	Types       []ResourceSubType  `json:"types,omitempty"`
	Immutable   bool               `json:"immutable,omitempty"`
	Abstract    bool               `json:"abstract,omitempty"`
	DataSource  *DataSource        `json:"dataSource,omitempty"`
	Entity      *string            `json:"entity,omitempty"`
	Catalog     *string            `json:"catalog,omitempty"`
	Title       *string            `json:"title,omitempty"`
	Description *string            `json:"description,omitempty"`
	Annotations map[string]string  `json:"annotations,omitempty"`
}

func (s *Resource) GetId() *uuid.UUID {
	return s.Id
}
func (s *Resource) GetVersion() int32 {
	return s.Version
}
func (s *Resource) GetCreatedBy() *string {
	return s.CreatedBy
}
func (s *Resource) GetUpdatedBy() *string {
	return s.UpdatedBy
}
func (s *Resource) GetCreatedOn() *time.Time {
	return s.CreatedOn
}
func (s *Resource) GetUpdatedOn() *time.Time {
	return s.UpdatedOn
}
func (s *Resource) GetName() string {
	return s.Name
}
func (s *Resource) GetNamespace() *Namespace {
	return s.Namespace
}
func (s *Resource) GetVirtual() bool {
	return s.Virtual
}
func (s *Resource) GetProperties() []ResourceProperty {
	return s.Properties
}
func (s *Resource) GetIndexes() []ResourceIndex {
	return s.Indexes
}
func (s *Resource) GetTypes() []ResourceSubType {
	return s.Types
}
func (s *Resource) GetImmutable() bool {
	return s.Immutable
}
func (s *Resource) GetAbstract() bool {
	return s.Abstract
}
func (s *Resource) GetDataSource() *DataSource {
	return s.DataSource
}
func (s *Resource) GetEntity() *string {
	return s.Entity
}
func (s *Resource) GetCatalog() *string {
	return s.Catalog
}
func (s *Resource) GetTitle() *string {
	return s.Title
}
func (s *Resource) GetDescription() *string {
	return s.Description
}
func (s *Resource) GetAnnotations() map[string]string {
	return s.Annotations
}

type ResourceProperty struct {
	Name         string                     `json:"name,omitempty"`
	Type         int32                      `json:"type,omitempty"`
	TypeRef      *string                    `json:"typeRef,omitempty"`
	Mapping      string                     `json:"mapping,omitempty"`
	Primary      bool                       `json:"primary,omitempty"`
	Required     bool                       `json:"required,omitempty"`
	Unique       bool                       `json:"unique,omitempty"`
	Immutable    bool                       `json:"immutable,omitempty"`
	Length       int32                      `json:"length,omitempty"`
	Item         *ResourceProperty          `json:"item,omitempty"`
	Properties   []ResourceProperty         `json:"properties,omitempty"`
	Reference    *ResourceReference         `json:"reference,omitempty"`
	DefaultValue *unstructured.Unstructured `json:"defaultValue,omitempty"`
	EnumValues   []string                   `json:"enumValues,omitempty"`
	ExampleValue *unstructured.Unstructured `json:"exampleValue,omitempty"`
	Title        *string                    `json:"title,omitempty"`
	Description  *string                    `json:"description,omitempty"`
	Annotations  map[string]string          `json:"annotations,omitempty"`
}

func (s *ResourceProperty) GetName() string {
	return s.Name
}
func (s *ResourceProperty) GetType() int32 {
	return s.Type
}
func (s *ResourceProperty) GetTypeRef() *string {
	return s.TypeRef
}
func (s *ResourceProperty) GetMapping() string {
	return s.Mapping
}
func (s *ResourceProperty) GetPrimary() bool {
	return s.Primary
}
func (s *ResourceProperty) GetRequired() bool {
	return s.Required
}
func (s *ResourceProperty) GetUnique() bool {
	return s.Unique
}
func (s *ResourceProperty) GetImmutable() bool {
	return s.Immutable
}
func (s *ResourceProperty) GetLength() int32 {
	return s.Length
}
func (s *ResourceProperty) GetItem() *ResourceProperty {
	return s.Item
}
func (s *ResourceProperty) GetProperties() []ResourceProperty {
	return s.Properties
}
func (s *ResourceProperty) GetReference() *ResourceReference {
	return s.Reference
}
func (s *ResourceProperty) GetDefaultValue() *unstructured.Unstructured {
	return s.DefaultValue
}
func (s *ResourceProperty) GetEnumValues() []string {
	return s.EnumValues
}
func (s *ResourceProperty) GetExampleValue() *unstructured.Unstructured {
	return s.ExampleValue
}
func (s *ResourceProperty) GetTitle() *string {
	return s.Title
}
func (s *ResourceProperty) GetDescription() *string {
	return s.Description
}
func (s *ResourceProperty) GetAnnotations() map[string]string {
	return s.Annotations
}

type ResourceSubType struct {
	Name       string             `json:"name,omitempty"`
	Properties []ResourceProperty `json:"properties,omitempty"`
}

func (s *ResourceSubType) GetName() string {
	return s.Name
}
func (s *ResourceSubType) GetProperties() []ResourceProperty {
	return s.Properties
}

type ResourceIndexProperty struct {
	Name  string         `json:"name,omitempty"`
	Order *ResourceOrder `json:"order,omitempty"`
}

func (s *ResourceIndexProperty) GetName() string {
	return s.Name
}
func (s *ResourceIndexProperty) GetOrder() *ResourceOrder {
	return s.Order
}

type ResourceIndex struct {
	Properties  []ResourceIndexProperty `json:"properties,omitempty"`
	IndexType   *ResourceIndexType      `json:"indexType,omitempty"`
	Unique      *bool                   `json:"unique,omitempty"`
	Annotations map[string]string       `json:"annotations,omitempty"`
}

func (s *ResourceIndex) GetProperties() []ResourceIndexProperty {
	return s.Properties
}
func (s *ResourceIndex) GetIndexType() *ResourceIndexType {
	return s.IndexType
}
func (s *ResourceIndex) GetUnique() *bool {
	return s.Unique
}
func (s *ResourceIndex) GetAnnotations() map[string]string {
	return s.Annotations
}

type ResourceReference struct {
	Resource      *Resource `json:"resource,omitempty"`
	Cascade       *bool     `json:"cascade,omitempty"`
	BackReference *string   `json:"backReference,omitempty"`
}

func (s *ResourceReference) GetResource() *Resource {
	return s.Resource
}
func (s *ResourceReference) GetCascade() *bool {
	return s.Cascade
}
func (s *ResourceReference) GetBackReference() *string {
	return s.BackReference
}

type ResourceOrder string

const (
	ResourceOrder_UNKNOWN ResourceOrder = "UNKNOWN"
	ResourceOrder_ASC     ResourceOrder = "ASC"
	ResourceOrder_DESC    ResourceOrder = "DESC"
)

type ResourceIndexType string

const (
	ResourceIndexType_BTREE ResourceIndexType = "BTREE"
	ResourceIndexType_HASH  ResourceIndexType = "HASH"
)

File: sdk/java/src/main/java/io/apibrew/lib/model/Resource.java
package io.apibrew.lib.model;

import java.util.Objects;
import io.apibrew.lib.EntityInfo;
import io.apibrew.lib.Entity;

public class Resource extends Entity {
    private java.util.UUID id;
    private int version;
    private String createdBy;
    private String updatedBy;
    private java.time.LocalDateTime createdOn;
    private java.time.LocalDateTime updatedOn;
    private String name;
    private Namespace namespace;
    private boolean virtual;
    private java.util.List<Resource.Property> properties;
    private java.util.List<Resource.Index> indexes;
    private java.util.List<Resource.SubType> types;
    private boolean immutable;
    private boolean $abstract;
    private boolean checkReferences;
    private DataSource dataSource;
    private String entity;
    private String catalog;
    private String title;
    private String description;
    private java.util.Map<String, String> annotations;

    public static final EntityInfo<Resource> entityInfo = new EntityInfo<>("system", "Resource", Resource.class);

    public static class Property {
        private String name;
        private Resource.Type type;
        private String typeRef;
        private String mapping;
        private boolean primary;
        private boolean required;
        private boolean unique;
        private boolean immutable;
        private int length;
        private Resource.Property item;
        private Resource.Reference reference;
        private Object defaultValue;
        private java.util.List<String> enumValues;
        private Object exampleValue;
        private String title;
        private String description;
        private java.util.Map<String, String> annotations;

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }
        public Resource.Type getType() {
            return type;
        }

        public void setType(Resource.Type type) {
            this.type = type;
        }
        public String getTypeRef() {
            return typeRef;
        }

        public void setTypeRef(String typeRef) {
            this.typeRef = typeRef;
        }
        public String getMapping() {
            return mapping;
        }

        public void setMapping(String mapping) {
            this.mapping = mapping;
        }
        public boolean getPrimary() {
            return primary;
        }

        public void setPrimary(boolean primary) {
            this.primary = primary;
        }
        public boolean getRequired() {
            return required;
        }

        public void setRequired(boolean required) {
            this.required = required;
        }
        public boolean getUnique() {
            return unique;
        }

        public void setUnique(boolean unique) {
            this.unique = unique;
        }
        public boolean getImmutable() {
            return immutable;
        }

        public void setImmutable(boolean immutable) {
            this.immutable = immutable;
        }
        public int getLength() {
            return length;
        }

        public void setLength(int length) {
            this.length = length;
        }
        public Resource.Property getItem() {
            return item;
        }

        public void setItem(Resource.Property item) {
            this.item = item;
        }
        public Resource.Reference getReference() {
            return reference;
        }

        public void setReference(Resource.Reference reference) {
            this.reference = reference;
        }
        public Object getDefaultValue() {
            return defaultValue;
        }

        public void setDefaultValue(Object defaultValue) {
            this.defaultValue = defaultValue;
        }
        public java.util.List<String> getEnumValues() {
            return enumValues;
        }

        public void setEnumValues(java.util.List<String> enumValues) {
            this.enumValues = enumValues;
        }
        public Object getExampleValue() {
            return exampleValue;
        }

        public void setExampleValue(Object exampleValue) {
            this.exampleValue = exampleValue;
        }
        public String getTitle() {
            return title;
        }

        public void setTitle(String title) {
            this.title = title;
        }
        public String getDescription() {
            return description;
        }

        public void setDescription(String description) {
            this.description = description;
        }
        public java.util.Map<String, String> getAnnotations() {
            return annotations;
        }

        public void setAnnotations(java.util.Map<String, String> annotations) {
            this.annotations = annotations;
        }
    }
    public static class SubType {
        private String name;
        private String title;
        private String description;
        private java.util.List<Resource.Property> properties;

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }
        public String getTitle() {
            return title;
        }

        public void setTitle(String title) {
            this.title = title;
        }
        public String getDescription() {
            return description;
        }

        public void setDescription(String description) {
            this.description = description;
        }
        public java.util.List<Resource.Property> getProperties() {
            return properties;
        }

        public void setProperties(java.util.List<Resource.Property> properties) {
            this.properties = properties;
        }
    }
    public static class IndexProperty {
        private String name;
        private Resource.Order order;

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }
        public Resource.Order getOrder() {
            return order;
        }

        public void setOrder(Resource.Order order) {
            this.order = order;
        }
    }
    public static class Index {
        private java.util.List<Resource.IndexProperty> properties;
        private Resource.IndexType indexType;
        private Boolean unique;
        private java.util.Map<String, String> annotations;

        public java.util.List<Resource.IndexProperty> getProperties() {
            return properties;
        }

        public void setProperties(java.util.List<Resource.IndexProperty> properties) {
            this.properties = properties;
        }
        public Resource.IndexType getIndexType() {
            return indexType;
        }

        public void setIndexType(Resource.IndexType indexType) {
            this.indexType = indexType;
        }
        public Boolean getUnique() {
            return unique;
        }

        public void setUnique(Boolean unique) {
            this.unique = unique;
        }
        public java.util.Map<String, String> getAnnotations() {
            return annotations;
        }

        public void setAnnotations(java.util.Map<String, String> annotations) {
            this.annotations = annotations;
        }
    }
    public static class Reference {
        private Resource resource;
        private Boolean cascade;
        private String backReference;

        public Resource getResource() {
            return resource;
        }

        public void setResource(Resource resource) {
            this.resource = resource;
        }
        public Boolean getCascade() {
            return cascade;
        }

        public void setCascade(Boolean cascade) {
            this.cascade = cascade;
        }
        public String getBackReference() {
            return backReference;
        }

        public void setBackReference(String backReference) {
            this.backReference = backReference;
        }
    }

    public static enum Type {
        BOOL("BOOL"),
        STRING("STRING"),
        FLOAT32("FLOAT32"),
        FLOAT64("FLOAT64"),
        INT32("INT32"),
        INT64("INT64"),
        BYTES("BYTES"),
        UUID("UUID"),
        DATE("DATE"),
        TIME("TIME"),
        TIMESTAMP("TIMESTAMP"),
        OBJECT("OBJECT"),
        MAP("MAP"),
        LIST("LIST"),
        REFERENCE("REFERENCE"),
        ENUM("ENUM"),
        STRUCT("STRUCT");

        private final String value;

        Type(String value) {
            this.value = value;
        }

        public String getValue() {
            return value;
        }
    }
    public static enum Order {
        UNKNOWN("UNKNOWN"),
        ASC("ASC"),
        DESC("DESC");

        private final String value;

        Order(String value) {
            this.value = value;
        }

        public String getValue() {
            return value;
        }
    }
    public static enum IndexType {
        BTREE("BTREE"),
        HASH("HASH");

        private final String value;

        IndexType(String value) {
            this.value = value;
        }

        public String getValue() {
            return value;
        }
    }

    public Resource() {
    }

    public EntityInfo<Resource> getEntityInfo() {
        return entityInfo;
    }

    public java.util.UUID getId() {
        return id;
    }

    public void setId(java.util.UUID id) {
        this.id = id;
    }
    public int getVersion() {
        return version;
    }

    public void setVersion(int version) {
        this.version = version;
    }
    public String getCreatedBy() {
        return createdBy;
    }

    public void setCreatedBy(String createdBy) {
        this.createdBy = createdBy;
    }
    public String getUpdatedBy() {
        return updatedBy;
    }

    public void setUpdatedBy(String updatedBy) {
        this.updatedBy = updatedBy;
    }
    public java.time.LocalDateTime getCreatedOn() {
        return createdOn;
    }

    public void setCreatedOn(java.time.LocalDateTime createdOn) {
        this.createdOn = createdOn;
    }
    public java.time.LocalDateTime getUpdatedOn() {
        return updatedOn;
    }

    public void setUpdatedOn(java.time.LocalDateTime updatedOn) {
        this.updatedOn = updatedOn;
    }
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
    public Namespace getNamespace() {
        return namespace;
    }

    public void setNamespace(Namespace namespace) {
        this.namespace = namespace;
    }
    public boolean getVirtual() {
        return virtual;
    }

    public void setVirtual(boolean virtual) {
        this.virtual = virtual;
    }
    public java.util.List<Resource.Property> getProperties() {
        return properties;
    }

    public void setProperties(java.util.List<Resource.Property> properties) {
        this.properties = properties;
    }
    public java.util.List<Resource.Index> getIndexes() {
        return indexes;
    }

    public void setIndexes(java.util.List<Resource.Index> indexes) {
        this.indexes = indexes;
    }
    public java.util.List<Resource.SubType> getTypes() {
        return types;
    }

    public void setTypes(java.util.List<Resource.SubType> types) {
        this.types = types;
    }
    public boolean getImmutable() {
        return immutable;
    }

    public void setImmutable(boolean immutable) {
        this.immutable = immutable;
    }
    public boolean getAbstract() {
        return $abstract;
    }

    public void setAbstract(boolean $abstract) {
        this.$abstract = $abstract;
    }
    public boolean getCheckReferences() {
        return checkReferences;
    }

    public void setCheckReferences(boolean checkReferences) {
        this.checkReferences = checkReferences;
    }
    public DataSource getDataSource() {
        return dataSource;
    }

    public void setDataSource(DataSource dataSource) {
        this.dataSource = dataSource;
    }
    public String getEntity() {
        return entity;
    }

    public void setEntity(String entity) {
        this.entity = entity;
    }
    public String getCatalog() {
        return catalog;
    }

    public void setCatalog(String catalog) {
        this.catalog = catalog;
    }
    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }
    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }
    public java.util.Map<String, String> getAnnotations() {
        return annotations;
    }

    public void setAnnotations(java.util.Map<String, String> annotations) {
        this.annotations = annotations;
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Resource)) {
            return false;
        }

        Resource obj = (Resource) o;

        if (!Objects.equals(this.id, obj.id)) {
            return false;
        }
        if (!Objects.equals(this.version, obj.version)) {
            return false;
        }
        if (!Objects.equals(this.createdBy, obj.createdBy)) {
            return false;
        }
        if (!Objects.equals(this.updatedBy, obj.updatedBy)) {
            return false;
        }
        if (!Objects.equals(this.createdOn, obj.createdOn)) {
            return false;
        }
        if (!Objects.equals(this.updatedOn, obj.updatedOn)) {
            return false;
        }
        if (!Objects.equals(this.name, obj.name)) {
            return false;
        }
        if (!Objects.equals(this.namespace, obj.namespace)) {
            return false;
        }
        if (!Objects.equals(this.virtual, obj.virtual)) {
            return false;
        }
        if (!Objects.equals(this.properties, obj.properties)) {
            return false;
        }
        if (!Objects.equals(this.indexes, obj.indexes)) {
            return false;
        }
        if (!Objects.equals(this.types, obj.types)) {
            return false;
        }
        if (!Objects.equals(this.immutable, obj.immutable)) {
            return false;
        }
        if (!Objects.equals(this.$abstract, obj.$abstract)) {
            return false;
        }
        if (!Objects.equals(this.checkReferences, obj.checkReferences)) {
            return false;
        }
        if (!Objects.equals(this.dataSource, obj.dataSource)) {
            return false;
        }
        if (!Objects.equals(this.entity, obj.entity)) {
            return false;
        }
        if (!Objects.equals(this.catalog, obj.catalog)) {
            return false;
        }
        if (!Objects.equals(this.title, obj.title)) {
            return false;
        }
        if (!Objects.equals(this.description, obj.description)) {
            return false;
        }
        if (!Objects.equals(this.annotations, obj.annotations)) {
            return false;
        }

        return true;
    }

    @Override
    public int hashCode() {
        if (id == null) {
            return super.hashCode();
        }

        return id.hashCode();
    }
}


